quiche/src/lib.rs

```diff
@@ -7757,18 +7761,28 @@ impl<F: BufFactory> Connection<F> {
                 let was_readable = stream.is_readable();
                 let priority_key = Arc::clone(&stream.priority_key);
 
-                let max_off_delta =
-                    stream.recv.reset(error_code, final_size)? as u64;
+                let stream::RecvBufResetReturn {
+                    max_data_delta,
+                    consumed_flowcontrol,
+                } = stream.recv.reset(error_code, final_size)?;
 
-                if max_off_delta > max_rx_data_left {
+                if max_data_delta > max_rx_data_left {
                     return Err(Error::FlowControl);
                 }
 
                 if !was_readable && stream.is_readable() {
                     self.streams.insert_readable(&priority_key);
                 }
 
-                self.rx_data += max_off_delta;
+                self.rx_data += max_data_delta;
+                // We dropped the recevice buffer, return connection level
+                // flow-control
+                self.flow_control.add_consumed(consumed_flowcontrol);
+
+                // ... and check if need to send an updated MAX_DATA frame
+                if self.should_update_max_data() {
+                    self.almost_full = true;
+                }
```

Comment:
From a code reuse perspective, there's a very similar block in the Shutdown::Read handler - https://github.com/cloudflare/quiche/pull/2158/files#diff-2ac678416f8d7ea462a714681b63acedcbbb7f2c99e537ce61b7667b3bf4ac41R5731. Maybe it would make sense to stick that in a function?

In fact, all non-test uses self.should_update_max_data()  just result in setting self.almost_full = true, so it seems a bit weird it is how is. But I can live with not changing it as part of this PR.

Reply:
Thats ok. Happy to close comment with no action.